### PR TITLE
feat(auth): add `onTokenCreated` option

### LINKED ISSUE #86: Callback that gets triggered on token creation / refresh
**What’s missing?**

When an access token is created (either from the OAuth `code` or by exchanging the `refreshToken`) during a request I want a way to retrieve the authentication payload to save it for later usage (e.g. in a database):
- `token`
- `refreshToken`
- `expiresAt`
- `refreshTokenExpiresAt`

Idea of implementation (https://github.com/octokit/octokit.js/discussions/2294#discussioncomment-3540412):
> `@octokit/auth-oauth-user` already has an `onVerification` option, we could add a new `onTokenCreated` option, which could pass a flag whether it was a refresh or not?

**Why?**

The library provides on the fly token refresh ([See code](https://github.com/octokit/auth-oauth-user.js/blob/main/src/auth.ts#L51-L70)), when providing the `refreshToken` and `expiresAt` like this:

```ts
import { createOAuthUserAuth } from '@octokit/auth-oauth-user';
import { graphql } from '@octokit/graphql';

const {
  accessToken,
  accessTokenExpiresAt,
  refreshToken,
  refreshTokenExpiresAt,
} = await getTokenFromDatabase();

const auth = createOAuthUserAuth({
  clientId: secretStore.getGithubClientId(),
  clientSecret: secretStore.getGithubClientSecret(),
  clientType: 'github-app',
  token: accessToken,
  expiresAt: accessTokenExpiresAt.toISOString(),
  refreshToken,
  refreshTokenExpiresAt: refreshTokenExpiresAt.toISOString(),
});

const apiClient = graphql.defaults({
  request: {
    hook: auth.hook,
  },
});
```

However it is impossible to detect whether the token was renewed during a request against the API this way and save the refreshed token back to the database.

Also see the discussion here for more information: https://github.com/octokit/octokit.js/discussions/2294

**Alternatives you tried**

--

### PR DESCRIPTION
**Pull request by oscard0m**
# Description
Introduce `onTokenCreated` option to `auth()` mehotd.

# Context
Fixes #86

### CODE REVIEW

**Comment by oscard0m** [comment]
This is a 5 min. draft PR to validate the approach. I would like your input here @gr2m @wolfy1339 @ofhouse

File: src/auth.ts [hunk 1]
```diff
@@ -102,6 +102,10 @@ export async function auth(
         ...authentication,
       };
 
+      // if (state.onTokenCreated) {
+      state.onTokenCreated(state.authentication, options.type);
```
Comment (oscard0m) [inline src/auth.ts:106]: Would this be the only place where we want to invoke the `onTokenCreated` or there are other places?

Comment (oscard0m) [inline src/auth.ts:106]: Passing `options.type` here would be enough info? @ofhouse

File: src/types.ts [hunk 2]
```diff
@@ -17,12 +17,14 @@ type CommonOAuthAppStrategyOptions = {
   clientId: string;
   clientSecret: string;
   request?: OctokitTypes.RequestInterface;
+  onTokenCreated?: Function;
```
Comment (oscard0m) [inline src/types.ts:20]: ℹ️ Temporary types for draft PR

File: src/auth.ts (hunk 1 above)
Comment (gr2m) [reply src/auth.ts:106]: only if `options.type` is `reset`. [`checkToken`](https://github.com/octokit/oauth-methods.js#checktoken) does not create a new token, it only checks if the current one is valid.

We also need to add the hook in `src/get-authentication.ts` at https://github.com/octokit/auth-oauth-user.js/blob/64558b5e44fd235ae40fe9529653da18f776fb60/src/get-authentication.ts#L39 and https://github.com/octokit/auth-oauth-user.js/blob/64558b5e44fd235ae40fe9529653da18f776fb60/src/get-authentication.ts#L59

Comment (gr2m) [reply src/auth.ts:106]: I'd prefer to pass an `{ type }` object, it will make it future proof in case we want to pass more options

**Review by gr2m — COMMENTED** [review]
approach looks good to me so far, just see my comments

**Comment by ofhouse** [comment]
@oscard0m Thank you very much for taking the initiative!

The existing `onVerification` callback seems to accept an async callback function.
For our case we will invoke a transaction on our database when the callback is triggered.
So it would be helpful if `onTokenCreated` would also optionally accept an async callback function.

File: src/types.ts (hunk 2 above)
Comment (oscard0m) [reply src/types.ts]: If somebody sees an opportunity to have better inferred types, I'm open to suggestions

File: src/auth.ts (hunk 1 above)
Comment (oscard0m) [reply src/auth.ts]: > only if `options.type` is `reset`. [`checkToken`](https://github.com/octokit/oauth-methods.js#checktoken) does not create a new token, it only checks if the current one is valid.

Covered here now: https://github.com/octokit/auth-oauth-user.js/pull/90/files#diff-84a5d130fb4cccb80f78601d140f1d5397dc0272f94cea672a8470539fcf6dc7R109-R114

> We also need to add the hook in `src/get-authentication.ts` at
> 
> https://github.com/octokit/auth-oauth-user.js/blob/64558b5e44fd235ae40fe9529653da18f776fb60/src/get-authentication.ts#L39
> 
> and
> https://github.com/octokit/auth-oauth-user.js/blob/64558b5e44fd235ae40fe9529653da18f776fb60/src/get-authentication.ts#L59

After checking the code, would it be simpler if we just call it at this point, after calling `getAuthentication()` a few lines above?

https://github.com/octokit/auth-oauth-user.js/pull/90/files#diff-84a5d130fb4cccb80f78601d140f1d5397dc0272f94cea672a8470539fcf6dc7R83-R86

**Comment by oscard0m** [comment]
> @oscard0m Thank you very much for taking the initiative!
> 
> The existing `onVerification` callback seems to accept an async callback function. For our case, we will invoke a transaction on our database when the callback is triggered. So it would be helpful if `onTokenCreated` would also optionally accept an async callback function.

Applied an `await` when invoking `onTokenCreated()` + added `void | Promise<void>` as return type

**Review by wolfy1339 — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -614,6 +614,7 @@ The differences are
 
 1. `scopes` is only present for OAuth Apps
 2. `refreshToken`, `expiresAt`, `refreshTokenExpiresAt` are only present for GitHub Apps, and only if token expiration is enabled
+3. `onTokenCreated` is only present for OAuth Apps and GitHub Apps
```
Comment (gr2m) [inline README.md:617]: you can remove that line, my understanding is the option behaves the same for both OAuth and GitHub Apps?

**Comment by coryvirok** [comment]
Looking forward to getting this merged and released.

File: README.md (hunk 3 above)
Comment (oscard0m) [reply README.md:617]:
```suggestion
```

**Review by wolfy1339 — APPROVED** [review]

**Pull request merged**